source/fundamentals/data-formats/codecs.txt

```diff
@@ -11,38 +11,337 @@ Codecs
+       object.
+
+The ``getEncoderClass()`` method returns a class instance of the Java class
+since Java cannot infer the type due to due to type erasure.
+
+.. _codecs-powerstatus-codec:
+
+See the following ``PowerStatusCodec`` class that implements ``Codec`` in
+order to convert the Java ``String`` values "on" and "off" to
+corresponding BSON boolean values.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/PowerStatusCodec.java
+   :start-after: start class
+   :end-before: end class
+   :language: java
+   :dedent:
+
+You can add an instance of your custom Codec to the ``CodecRegistry`` which
+contains a mapping between your ``Codec`` and the Java object type to which
+it applies. Continue to the :ref:`CodecRegistry <codecs-codecregistry>`
+section of this page to see how you can include your Codec.
+
+For more information about the classes and interfaces in this section, see the
+following API documentation:
+
+- :java-docs:`Codec <apidocs/bson/org/bson/codecs/Codec.html>`
+- :java-docs:`BsonWriter <apidocs/bson/org/bson/BsonWriter.html>`
+- :java-docs:`BsonBinaryWriter <apidocs/bson/org/bson/BsonBinaryWriter.html>`
+- :java-docs:`EncoderContext <apidocs/bson/org/bson/codecs/EncoderContext.html>`
+- :java-docs:`BsonReader <apidocs/bson/org/bson/BsonReader.html>`
+- :java-docs:`DecoderContext <apidocs/bson/org/bson/codecs/DecoderContext.html>`
+- :java-docs:`BsonBinaryReader <apidocs/bson/org/bson/BsonBinaryReader.html>`
+
+.. _codecs-codecregistry:
 
 CodecRegistry
 -------------
 
-- Description
-- Example of setting Codec list
-- Example of getting a Codec
+A ``CodecRegistry`` is an immutable collection of ``Codec`` instances that
+encode and decode the Java classes they specify. You can use any of the
+following ``CodecRegistries`` class static factory methods to construct a
+``CodecRegistry`` from the ``Codec`` instances contained in the associated
+types:
+
+- ``fromCodecs()``
+- ``fromProviders()``
+- ``fromRegistries()``
+
+The following code snippet shows how to construct a ``CodecRegistry`` using
+the ``fromCodecs()`` method:
+
+.. code-block:: java
+   :copyable: true
+
+   CodecRegistry codecRegistry = CodecRegistries.fromCodecs(new IntegerCodec(), new PowerStatusCodec());
+
+In the example above, we assign the ``CodecRegistry`` the following ``Codec``
+implementations:
+
+- :java-docs:`IntegerCodec </apidocs/bson/org/bson/codecs/IntegerCodec.html>`,
+  a ``Codec`` that converts Integers and is part of the BSON package.
+- :ref:`PowerStatusCodec <codecs-powerstatus-codec>`, our sample ``Codec``
+  that converts certain Java strings to BSON booleans.
+
+You can retrieve the Codecs from the ``CodecRegistry`` instance from the
+prior example using the following code:
+
+.. code-block:: java
+   :copyable: true
+
+   Codec<String> powerStatus = codecRegistry.get(String.class);
+   Codec<Integer> integerCodec = codecRegistry.get(Integer.class);
+
+If you attempt to retrieve a ``Codec`` instance for a class that is not
```

Comment:
I think this is should be in a note block to make it stand out more to the reader.

Reply:
Perhaps a `.. warning`

Reply:
Maybe I'm missing something about how we should use admonitions (note, warning, etc.), but here's my decision process:

I've only used warning or note blocks for describing some unexpected behavior whether it's a difference between how the driver and server behave or a common user mistake.

This specific exception type is a feature that helps a user identify what went wrong during runtime. I'm not sure we should be using an admonition to let a user know about this.

Let me know if this makes sense and/or whether I'm missing something.

Reply:
I think this makes sense, especially since it isn't surprising.